cmd/kubeadm/app/master/selfhosted.go

```diff
@@ -41,45 +42,81 @@ var (
 )
 
 func CreateSelfHostedControlPlane(cfg *kubeadmapi.MasterConfiguration, client *clientset.Clientset) error {
-	volumes := []v1.Volume{k8sVolume(cfg)}
-	volumeMounts := []v1.VolumeMount{k8sVolumeMount()}
-	if isCertsVolumeMountNeeded() {
-		volumes = append(volumes, certsVolume(cfg))
-		volumeMounts = append(volumeMounts, certsVolumeMount())
+	// Launch the API server
+	if err := createTLSSecrets(client); err != nil {
+		return err
 	}
 
-	if isPkiVolumeMountNeeded() {
-		volumes = append(volumes, pkiVolume(cfg))
-		volumeMounts = append(volumeMounts, pkiVolumeMount())
+	if err := launchSelfHostedAPIServer(cfg, client); err != nil {
+		return err
 	}
 
-	// Need lock for self-hosted
-	volumes = append(volumes, flockVolume())
-	volumeMounts = append(volumeMounts, flockVolumeMount())
+	wait.PollInfinite(kubeadmconstants.APICallRetryInterval, func() (bool, error) {
```

Comment:
I really think this should be a standalone, testable function. Also I think this logic exists somewhere in apiclient.go or so already